### DISCUSSION TITLE
Update Data

### DISCUSSION THREAD

**Discussion by mattjvincent** [genome-spy/genome-spy#77] [Ideas]
Is there a way to dynamically update the underlying data and redraw the visualization?

For example, at the following URL if you click the LOD Peaks tab, the plot will update while changing the slider.

[https://churchilllab.jax.org/qtlviewer/JAC/DOHeart](https://churchilllab.jax.org/qtlviewer/JAC/DOHeart)

Thoughts?

**Comment by tuner** [comment]
Not currently. There are a few things that must be solved first.

- [ ] Design the API. Perhaps datasources could be given names, and then: `api.publishData("genes", [{...}, {...}, ...])`. Vega has a nice API for that: https://vega.github.io/vega/docs/api/view/#view_data
- [ ] Get rid of the [optimization hack](https://github.com/genome-spy/genome-spy/blob/2abd5a2fb3e752bf651a7ff8fa4017df58fc12ab/packages/core/src/marks/mark.js#L741) that requires that all dynamically updated marks must have `dynamicData: true` in their configs. 
- [ ] Some minor fixes here and there to make it robust.

Internally GenomeSpy already updates data dynamically in several places such as on the Gene annotation track, axis ticks and tick labels, etc. There's still some work to do in order to support dynamic updates through an API. Definitely something I'd like to do, but cannot use time for this right now.

**Comment by tuner** [comment]
I have forgot to update this thread.

Lazy data sources have been implemented: https://genomespy.app/docs/grammar/data/lazy/

Some data sources can now be parameterized, but it's still a bit work in progress and yet to be documented properly: https://genomespy.app/docs/grammar/parameters/

And there's also an API for pushing new data: https://github.com/genome-spy/genome-spy/blob/master/packages/embed-examples/src/dynamicNamedData.js
